### PR TITLE
feat: get data for account details page

### PR DESCRIPTION
**Pull request by bkioshn**
## Describe your changes
- Refactor queries related file and folder
- Get data for account details page

Including:
1. Assets `useUserAssetInfos`
 - Supported Assets  (value sorted from highest to lowest)
 - Unsupported Assets

2. Transactions `useTxQuery`
- Use the same query as `PastTx` page

3. Contract instances `useContractInstances`

4. Contract Admins `useContractsAdmin`

5. Stored Code `useCodeStored`

6. Opened Proposal `useProposalsByWalletAddressPagination`

7. Data counts `useAccountDetailsTableCounts`
- Including the number of data for each table

### CODE REVIEW

File: src/lib/query/other.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { graphql } from "lib/gql";
+
+export const getBlockTimestampByHeightQueryDocument = graphql(`
```
Comment (poomthiti) [inline src/lib/query/other.ts:3]: queries/block.ts

File: src/lib/query/proposal.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      where: { contract: { address: { _eq: $contractAddress } } }
+    ) {
+      aggregate {
+        count
+      }
+    }
+  }
+`);
+
+export const getProposalsByUserAddress = graphql(`
```
Comment (poomthiti) [inline src/lib/query/proposal.ts:43]: WalletAddress instead of UserAddress

File: src/lib/services/accountService.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      currentChainRecord?.name,
+      currentChainRecord?.chain.chain_id,
+    ],
+    async () =>
+      getAccountBalanceInfo(
+        address,
+        currentChainRecord?.name,
+        currentChainRecord?.chain.chain_id
+      ),
+    { enabled: !!currentChainRecord }
```
Comment (poomthiti) [inline src/lib/services/accountService.ts:27]: address also

File: src/lib/services/accountService.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { useWallet } from "@cosmos-kit/react";
+import type { UseQueryResult } from "@tanstack/react-query";
+import { useQuery } from "@tanstack/react-query";
+
+import type { Balance, HumanAddr } from "lib/types";
+
+import { getAccountBalanceInfo } from "./account";
+
+export const useAccountBalance = (
+  address: HumanAddr
```
Comment (poomthiti) [inline src/lib/services/accountService.ts:10]: should be option

File: src/lib/pages/account-details/data.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+  const { data: contractsAdmin, isLoading } =
+    useContractListByAdminWithPagination(
+      address as HumanAddr,
+      offset,
+      pageSize
+    );
+  const { getContractLocalInfo } = useContractStore();
+
+  // eslint-disable-next-line sonarjs/no-identical-functions
```
Comment (poomthiti) [inline src/lib/pages/account-details/data.ts:50]: just change some variable name, maybe (contract) to (contractAdmin)

File: src/lib/services/proposalService.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      pageSize,
+    ],
+    queryFn,
+    {
+      enabled: !!address,
+    }
+  );
+};
+
+export const useProposalsCountByUserAddress = (
```
Comment (poomthiti) [inline src/lib/services/proposalService.ts:142]: WalletAddress

File: src/lib/services/contractService.ts [hunk 7]
```diff
@@ -440,72 +442,224 @@ export const useTxsCountByContractAddress = (
...
+      indexerGraphClient
+        .request(getContractListByWalletAddressWithPagination, {
+          walletAddress,
+          offset,
+          pageSize,
+        })
+        // eslint-disable-next-line sonarjs/no-identical-functions
+        .then(({ contracts }) =>
+          // eslint-disable-next-line sonarjs/no-identical-functions
+          contracts.map<ContractInfo>((contract) => ({
```
Comment (poomthiti) [inline src/lib/services/contractService.ts:530]: change mapping var name and remove eslint

**Review by poomthiti — CHANGES_REQUESTED** [review]

File: src/lib/pages/account-details/data.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+          .sub(
+            assetValue(
+              formatTokenWithPrecision(
+                a.balance.amount as Token,
+                a.balance.precision
+              ),
+              a.balance.price
+            )
+          )
+          .toNumber();
```
Comment (poomthiti) [inline src/lib/pages/account-details/data.ts:124]: use .toFixed()

**Review by poomthiti — CHANGES_REQUESTED** [review]

**Comment by poomthiti** [comment]
Also rename the remaining queries that have pagination

File: src/lib/pages/account-details/data.ts (hunk 8 above)
Comment (bkioshn) [reply src/lib/pages/account-details/data.ts:124]: Why toFixed() ?

Comment (poomthiti) [reply src/lib/pages/account-details/data.ts:122]: sorry my bad

File: src/lib/pages/account-details/data.ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import {
+  useContractListByAdminPagination,
+  useContractListByWalletAddressPagination,
+} from "lib/services/contractService";
+import type { BalanceWithAssetInfo, HumanAddr, Token } from "lib/types";
+import { formatTokenWithPrecision } from "lib/utils";
+import { assetValue } from "lib/utils/assetValue";
+
+export const useContractInstances = (
+  walletAddress: HumanAddr,
```
Comment (poomthiti) [inline src/lib/pages/account-details/data.ts:14]: Should be option

File: src/lib/pages/account-details/data.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+    return {
+      ...contractAdmin,
+      tags: localInfo?.tags,
+      contractName: localInfo?.name,
+    };
+  });
+
+  return {
+    contractsAdmin: data,
```
Comment (poomthiti) [inline src/lib/pages/account-details/data.ts:59]: you can use `contracts` as key here

File: src/lib/pages/account-details/data.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  };
+};
+
+export const useUserAssetInfos = () => {
+  const { address } = useWallet();
+
+  const { data: assets, isLoading } = useAccountBalance(address as HumanAddr);
+  const assetInfos = useAssetInfos();
+
+  const contractBalancesWithAssetInfos = assets?.map(
```
[the remark this answers is not in the document]
Comment (poomthiti) [reply src/lib/pages/account-details/data.ts:94]: please change var name

File: src/lib/services/proposalService.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    queryFn,
+    {
+      keepPreviousData: true,
+      enabled: !!contractAddress,
+    }
+  );
+};
+
+export const useProposalsByWalletAddressPagination = (
+  walletAddress: HumanAddr,
```
Comment (poomthiti) [inline src/lib/services/proposalService.ts:99]: Option

File: src/lib/utils/assetValue.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+import type { BigSource } from "big.js";
+import big from "big.js";
+
+export const assetValue = (amount: BigSource, price: BigSource) =>
```
Comment (poomthiti) [inline src/lib/utils/assetValue.ts:4]: calculateAssetValue

**Review by poomthiti — CHANGES_REQUESTED** [review]

File: src/lib/services/account.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  chainName: Option<string>,
+  chainId: Option<string>
+): Promise<Balance[]> => {
+  if (!chainName || !chainId)
+    throw new Error("Invalid chain (getAccountBalanceInfo)");
+
+  if (!walletAddr) {
+    throw new Error("Wallet address not found (getAccountBalanceInfo)");
+  }
+  const { data } = await axios.get(
```
Comment (evilpeach) [inline src/lib/services/account.ts:17]: Also add generic type of `axios.get` function like this `axios.get<Balance[]>`

File: src/lib/services/codeService.ts [hunk 15]
```diff
@@ -155,66 +153,72 @@ export const useCodeInfoByCodeId = (
...
-            contract.contract_histories.at(0)?.block.timestamp
-          ),
-          remark: contract.contract_histories.at(0)?.remark,
+      .request(getCodeListByWalletAddressPagination, {
+        walletAddress,
+        offset,
+        pageSize,
+      })
+      .then(({ codes }) =>
+        codes.map((code) => ({
```
Comment (evilpeach) [inline src/lib/services/codeService.ts:174]: ` codes.map<CodeInfo>((code) => ({`

File: src/lib/query/block.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import { graphql } from "lib/gql";
```
Comment (evilpeach) [inline src/lib/query/block.ts:1]: Create `index.ts` in `query` folder like this

```tsx
export * from './block.ts';
export * from './code.ts';
...
```

File: src/lib/services/contractService.ts [hunk 17]
```diff
@@ -350,163 +346,221 @@ export const useMigrationHistoriesCountByContractAddress = (
...
+    return indexerGraphClient
+      .request(getContractListCountByWalletAddress, {
+        walletAddress,
+      })
+      .then(({ contracts_aggregate }) => contracts_aggregate?.aggregate?.count);
+  }, [walletAddress, indexerGraphClient]);
+
+  return useQuery(
+    [
+      "contract_list_count_by_wallet_address_pagination",
```
Comment (evilpeach) [inline src/lib/services/contractService.ts:485]: remove `_pagination`?

File: src/lib/services/proposalService.ts (hunk 12 above)
Comment (evilpeach) [inline src/lib/services/proposalService.ts:99]: Option?

Comment (evilpeach) [inline src/lib/services/proposalService.ts:99]: Is it optional or not?

File: src/lib/services/proposalService.ts [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+      );
+
+    return indexerGraphClient
+      .request(getProposalsByWalletAddressPagination, {
+        walletAddress: walletAddress as HumanAddr,
+        offset,
+        pageSize,
+      })
+      .then(({ proposals }) =>
+        proposals.map((proposal) => ({
```
Comment (evilpeach) [inline src/lib/services/proposalService.ts:117]: `proposals.map<UserProposal>`

File: src/lib/services/proposalService.ts [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+      enabled: !!contractAddress,
+    }
+  );
+};
+
+export const useProposalsByWalletAddressPagination = (
+  walletAddress: HumanAddr,
+  offset: number,
+  pageSize: number
+): UseQueryResult<UserProposal> => {
```
Comment (evilpeach) [inline src/lib/services/proposalService.ts:102]: `<UserProposal[]>`

File: src/lib/services/proposalService.ts [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+    ],
+    queryFn,
+    {
+      enabled: !!walletAddress,
+    }
+  );
+};
+
+export const useProposalsCountByWalletAddress = (
+  walletAddress: Option<HumanAddr>
```
Comment (evilpeach) [inline src/lib/services/proposalService.ts:144]: Is it optional or not? It is not the same as above.

File: src/lib/services/txService.ts [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+} from "lib/query/tx";
+import type { Addr, AllTransaction, ContractAddr, Option } from "lib/types";
+import {
+  getActionMsgType,
+  parseDateOpt,
+  parseTxHash,
+  snakeToCamel,
+  unwrapAll,
+} from "lib/utils";
+
```
Comment (evilpeach) [inline src/lib/services/txService.ts:18]: How about getTxByWalletAddress`?

File: src/lib/pages/account-details/data.ts [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+import { useAccountBalances } from "lib/services/accountService";
+import { useAssetInfos } from "lib/services/assetService";
+import { useCodeListByWalletAddressPagination } from "lib/services/codeService";
+import {
+  useContractListByAdminPagination,
+  useContractListByWalletAddressPagination,
+} from "lib/services/contractService";
+import type { BalanceWithAssetInfo, HumanAddr, Token, Option } from "lib/types";
+import { formatTokenWithPrecision } from "lib/utils";
+import { calculateAssetValue } from "lib/utils/assetValue";
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:11]: merge

Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:11]: import `calculateAssetValue` in line 10

File: src/lib/pages/account-details/data.ts [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+
+export const useContractInstances = (
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:13]: declare return interface (singlar form)

File: src/lib/pages/account-details/data.ts [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+  walletAddress: Option<HumanAddr>,
+  offset: number,
+  pageSize: number
+) => {
+  const { data: contracts, isLoading } =
+    useContractListByWalletAddressPagination(walletAddress, offset, pageSize);
+  const { getContractLocalInfo } = useContractStore();
+  const data = contracts?.map((contract) => {
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:21]: declare interface for return data

File: src/lib/pages/account-details/data.ts [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    contracts: data,
+    isLoading,
+  };
+};
+
+export const useContractsAdmin = (
+  walletAddress: Option<HumanAddr>,
+  offset: number,
+  pageSize: number
+) => {
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:40]: declare and use data interface of output

File: src/lib/pages/account-details/data.ts [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+    codes: data,
+    isLoading,
+  };
+};
+
+export const useUserAssetInfos = (walletAddress: Option<HumanAddr>) => {
+  const { data: balances, isLoading } = useAccountBalances(walletAddress);
+  const assetInfos = useAssetInfos();
+
+  const contractBalancesWithAssetInfos = balances?.map(
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:94]: `map<BalanceWithAssetInfo>`

File: src/lib/model/account.ts [hunk 27]
```diff
@@ -0,0 +1,54 @@
+import { useTxQueryCount } from "lib/pages/past-txs/query/useTxQuery";
```
Comment (evilpeach) [inline src/lib/model/account.ts:1]: add `todo` refactor later

File: src/lib/model/account.ts [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+} from "lib/services/contractService";
+import { useProposalsCountByWalletAddress } from "lib/services/proposalService";
+import type { HumanAddr, Option } from "lib/types";
+
+/**
+ * @remark
+ * Counts for stored codes, contract admin, contract instances, transactions, and opened proposals tables
+ */
+export const useAccountDetailsTableCounts = (
+  walletAddress: Option<HumanAddr>
```
Comment (evilpeach) [inline src/lib/model/account.ts:15]: I think we can make sure `walletAddress` can't be optional in UI section.

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/query/contract.ts [hunk 29]
```diff
@@ -314,78 +262,63 @@ export const getContractListByCodeId = graphql(`
   }
 `);
 
-export const getContractListCountByCodeId = graphql(`
-  query getContractListCountByCodeId($codeId: Int!) {
-    contracts_aggregate(where: { code_id: { _eq: $codeId } }) {
+export const getContractListCountByWalletAddress = graphql(`
```
[the remark this answers is not in the document]
Comment (bkioshn) [reply src/lib/query/contract.ts:265]: Will use `getInstantiatedCountByUserQueryDocument` instead

File: src/lib/services/proposalService.ts (hunk 20 above)
Comment (bkioshn) [reply src/lib/services/proposalService.ts:144]: Should not be optional

File: src/lib/services/txService.ts (hunk 21 above)
Comment (bkioshn) [reply src/lib/services/txService.ts:18]: Transaciton will use same query as past tx page

File: src/lib/services/accountService.ts [hunk 30]
```diff
@@ -0,0 +1,29 @@
+import { useWallet } from "@cosmos-kit/react";
+import type { UseQueryResult } from "@tanstack/react-query";
+import { useQuery } from "@tanstack/react-query";
+
+import type { Balance, HumanAddr, Option } from "lib/types";
+
+import { getAccountBalanceInfo } from "./account";
+
+export const useAccountBalances = (
+  walletAddress: Option<HumanAddr>
```
Comment (evilpeach) [inline src/lib/services/accountService.ts:10]: remove option

File: src/lib/services/account.ts [hunk 31]
```diff
@@ -0,0 +1,23 @@
+import axios from "axios";
+
+import { CELATONE_API_ENDPOINT, getChainApiPath } from "env";
+import type { Balance, HumanAddr, Option } from "lib/types";
+
+export const getAccountBalanceInfo = async (
+  walletAddr: Option<HumanAddr>,
```
Comment (evilpeach) [inline src/lib/services/account.ts:7]: remove option

File: src/lib/services/contractService.ts [hunk 32]
```diff
@@ -452,7 +451,7 @@ export const useContractListByWalletAddressPagination = (
 
   return useQuery(
     [
-      "contract_list_by_wallet_address_pagination",
+      "contract_list_by_wallet_address",
```
Comment (evilpeach) [inline src/lib/services/contractService.ts:454]: contract_list_by_wallet_address_pagination

File: src/lib/pages/account-details/data.ts [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+                a.balance.amount as Token,
+                a.balance.precision
+              ),
+              a.balance.price
+            )
+          )
+          .toNumber();
+      }
+      return 1;
+    });
```
Comment (evilpeach) [inline src/lib/pages/account-details/data.ts:149]: I suggest to do this
```tsx
.sort((a, b) => a.balance.price && b.balance.price ? calAssetValueByPrecision(b).cmp(calAssetValuePrecision(a)) : 1;
})

```

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by evilpeach — APPROVED** [review]

**Review by songwongtp — APPROVED** [review]

**Pull request merged**